crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -0,0 +1,274 @@
+        (_, Type::BooleanLiteral(_)) => Ordering::Greater,
+
+        (Type::IntLiteral(left), Type::IntLiteral(right)) => left.cmp(right),
+        (Type::IntLiteral(_), _) => Ordering::Less,
+        (_, Type::IntLiteral(_)) => Ordering::Greater,
+
+        (Type::StringLiteral(left), Type::StringLiteral(right)) => left.cmp(right),
+        (Type::StringLiteral(_), _) => Ordering::Less,
+        (_, Type::StringLiteral(_)) => Ordering::Greater,
+
+        (Type::BytesLiteral(left), Type::BytesLiteral(right)) => left.cmp(right),
+        (Type::BytesLiteral(_), _) => Ordering::Less,
+        (_, Type::BytesLiteral(_)) => Ordering::Greater,
+
+        (Type::SliceLiteral(left), Type::SliceLiteral(right)) => left.cmp(right),
+        (Type::SliceLiteral(_), _) => Ordering::Less,
+        (_, Type::SliceLiteral(_)) => Ordering::Greater,
+
+        (Type::FunctionLiteral(left), Type::FunctionLiteral(right)) => left.cmp(right),
+        (Type::FunctionLiteral(_), _) => Ordering::Less,
+        (_, Type::FunctionLiteral(_)) => Ordering::Greater,
+
+        (Type::Tuple(left), Type::Tuple(right)) => left.cmp(right),
+        (Type::Tuple(_), _) => Ordering::Less,
+        (_, Type::Tuple(_)) => Ordering::Greater,
+
+        (Type::ModuleLiteral(left), Type::ModuleLiteral(right)) => left.cmp(right),
+        (Type::ModuleLiteral(_), _) => Ordering::Less,
+        (_, Type::ModuleLiteral(_)) => Ordering::Greater,
+
+        (
+            Type::ClassLiteral(ClassLiteralType { class: left }),
+            Type::ClassLiteral(ClassLiteralType { class: right }),
+        ) => left.cmp(right),
+        (Type::ClassLiteral(_), _) => Ordering::Less,
+        (_, Type::ClassLiteral(_)) => Ordering::Greater,
+
+        (Type::SubclassOf(left), Type::SubclassOf(right)) => {
+            match (left.subclass_of(), right.subclass_of()) {
+                (ClassBase::Class(left), ClassBase::Class(right)) => left.cmp(&right),
+                (ClassBase::Class(_), _) => Ordering::Less,
+                (_, ClassBase::Class(_)) => Ordering::Greater,
+                (ClassBase::Dynamic(left), ClassBase::Dynamic(right)) => {
+                    order_dynamic_elements(left, right)
+                }
+            }
+        }
+
+        (Type::SubclassOf(_), _) => Ordering::Less,
+        (_, Type::SubclassOf(_)) => Ordering::Greater,
+        (
+            Type::Instance(InstanceType { class: left }),
+            Type::Instance(InstanceType { class: right }),
+        ) => left.cmp(right),
+
+        (Type::Instance(_), _) => Ordering::Less,
+        (_, Type::Instance(_)) => Ordering::Greater,
+
+        (Type::AlwaysTruthy, _) => Ordering::Less,
+        (_, Type::AlwaysTruthy) => Ordering::Greater,
+
+        (Type::AlwaysFalsy, _) => Ordering::Less,
+        (_, Type::AlwaysFalsy) => Ordering::Greater,
+
+        (Type::KnownInstance(left_instance), Type::KnownInstance(right_instance)) => {
+            match (left_instance, right_instance) {
```

Comment:
I think implementing `PartialOrd` for `KnownInstance` should be fine. I agree that `Type` itself should not implement `Ord`. It would drastically reduce the code needed here (and I'd probably to the same for other types, e.g. `ClassBase)

Reply:
> I think implementing `PartialOrd` for `KnownInstance` should be fine.

Hmm, why do you think it's okay for `KnownInstanceType` but not for `Type`? The two enums seem pretty similar to me in terms of whether it makes sense for them to implement `Ord`

Reply:
I consider `Type` as the main public interface. I agree that probably the same argument applies as for `Type`.

Reply:
I kinda feel inclined to add it to all of these heterogeneous enums or none of them. Writing it all out like this is a lot of boilerplate code, but it's not _too_ hard to review it for correctness (it's all quite simple). But maybe it _would_ be better to just write a macro here...